torch/_inductor/runtime/triton_bundler.py

```diff
@@ -0,0 +1,127 @@
+import dataclasses
+import logging
+import os
+from pathlib import Path
+from typing import Dict, List, Optional
+
+import torch
+from torch._dynamo.utils import counters
+
+from .runtime_utils import cache_dir
+
+
+log = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass(frozen=True)
+class TritonBundleEntry:
+    kernel_hash: str
+    device: int
+    directory: str
+
+
+@dataclasses.dataclass
+class TritonKernelArtifact:
+    filename: str
+    payload: bytes = dataclasses.field(repr=False)  # Do not display binary
+
+
+@dataclasses.dataclass
+class TritonKernelArtifacts:
+    kernel_hash: str
+    device: int
+    artifacts: List[TritonKernelArtifact]
+
+
+class TritonBundler:
+    # It is safe for multiple threads to insert entries as insert operation
+    # of dict is atomic
+    _entries: Optional[Dict[TritonBundleEntry, None]] = None
+
+    @staticmethod
+    def is_enabled() -> bool:
+        from torch._inductor import config
+
+        if (b := config.bundle_triton_into_fx_graph_cache) is not None:
+            return b
+
+        if not config.is_fbcode():
+            return False
+
+        return torch._utils_internal.justknobs_check(
+            "pytorch/remote_cache:bundle_triton_into_fx_graph_cache"
+        )
+
+    @classmethod
+    def begin_compile(cls) -> None:
+        if not TritonBundler.is_enabled():
+            return
+        assert cls._entries is None
+        cls._entries = {}
+
+    @classmethod
+    def put(cls, kernel_hash: str, device: int) -> None:
+        if (entries := cls._entries) is not None:
+            # CachingAutotuner.__init__ unconditionally sets TRITON_CACHE_DIR
+            triton_cache_dir = os.getenv("TRITON_CACHE_DIR")
+            assert triton_cache_dir is not None
+            entries[TritonBundleEntry(kernel_hash, device, triton_cache_dir)] = None
+
+    @classmethod
+    def collect(cls) -> List[TritonKernelArtifacts]:
+        if not TritonBundler.is_enabled():
+            cls._entries = None
+            return []
+
+        if (entries := cls._entries) is not None:
+            result: List[TritonKernelArtifacts] = []
+            for entry in entries.keys():
+                artifacts: List[TritonKernelArtifact] = []
+                path = os.path.join(entry.directory, entry.kernel_hash)
+                if not os.path.exists(path):
+                    continue
+                for filename in os.listdir(path):
+                    filepath = os.path.join(path, filename)
+                    try:
+                        assert os.path.isfile(filepath)
+                        with open(filepath, "rb") as file:
+                            artifacts.append(
+                                TritonKernelArtifact(filename, file.read())
```

Comment:
qq - Is there a reason we don't want to keep the triton items in memory? Reading them off the file system is fine, but it does introduce more points of failure (i.e.disk space is full, multiple processes accidentally writing at the same time, etc...).

Reply:
We still have to rely on the triton compiler, so this is a way to feed it off of to the triton compiler. We do not have a deep integration with the compiler to feed it in memory objects

Reply:
sgtm, maybe mention that in a comment?